rskj-core/src/test/java/co/rsk/peg/union/UnionStorageProviderImplTest.java

```diff
@@ -0,0 +1,187 @@
+    void getUnionAddress_whenNoAddressStoredOrSet_shouldReturnEmpty() {
+        // Act
+        Optional<RskAddress> unionBridgeAddress = unionStorageProvider.getUnionAddress();
+
+        // Assert
+        assertTrue(unionBridgeAddress.isEmpty());
+    }
+
+    private static Stream<Arguments> validAddressProvider() {
+        return Stream.of(
+            Arguments.of(new RskAddress(new byte[20])), // empty address: 0000000000000000000000000000000000
+            Arguments.of(TestUtils.generateAddress("address"))
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("validAddressProvider")
+    void getUnionAddress_whenAddressStored_shouldReturnStoredAddress(RskAddress rskAddress) {
+        // Arrange
+        storageAccessor.saveToRepository(
+            UnionStorageIndexKey.UNION_BRIDGE_CONTRACT_ADDRESS.getKey(),
+            rskAddress,
+            BridgeSerializationUtils::serializeRskAddress
+        );
+
+        // Act
+        Optional<RskAddress> unionBridgeAddress = unionStorageProvider.getUnionAddress();
+
+        // Assert
+        assertTrue(unionBridgeAddress.isPresent());
+        assertEquals(rskAddress, unionBridgeAddress.get());
+    }
+
+    @ParameterizedTest
+    @MethodSource("validAddressProvider")
+    void getUnionAddress_whenAddressSet_shouldReturnCacheAddress(RskAddress rskAddress) {
+        // Arrange
+        unionStorageProvider.setUnionAddress(rskAddress);
+
+        // Act
+        Optional<RskAddress> unionBridgeAddress = unionStorageProvider.getUnionAddress();
+
+        // Assert
+        assertTrue(unionBridgeAddress.isPresent());
+        assertEquals(rskAddress, unionBridgeAddress.get());
+    }
+
+    @ParameterizedTest
+    @MethodSource("validAddressProvider")
+    void getUnionAddress_whenAddressSetAndStored_shouldReturnSetAddress(RskAddress rskAddress) {
+        // Arrange
+        unionStorageProvider.setUnionAddress(rskAddress);
+        unionStorageProvider.save();
+
+        // Act
+        Optional<RskAddress> unionBridgeAddress = unionStorageProvider.getUnionAddress();
+
+        // Assert
+        assertTrue(unionBridgeAddress.isPresent());
+        assertEquals(rskAddress, unionBridgeAddress.get());
+    }
+
+    @Test
+    void setUnionAddress_whenAddressSet_shouldNotStore() {
+        // Arrange
+        RskAddress rskAddress = TestUtils.generateAddress("address");
+
+        // Act
+        unionStorageProvider.setUnionAddress(rskAddress);
+
+        // Assert
+        verify(mockStorageAccessor, never())
+            .saveToRepository(any(), any(), any());
+    }
+
+    @Test
+    void save_whenNoAddressSet_shouldNotStoreAnything() {
+        // Act
+        unionStorageProviderUsingMockStorage.save();
+
+        // Assert
+        verify(mockStorageAccessor, never())
+            .saveToRepository(any(), any(), any());
+    }
+
+    @ParameterizedTest
+    @MethodSource("validAddressProvider")
+    void getAndSetUnionAddress_whenFirstTime_ok(RskAddress rskAddress) {
```

Comment:
```suggestion
    void getSetAndSaveUnionAddress_whenNoValueIsSet_shouldSaveNewAddress(RskAddress rskAddress) {
```
something like this a bit more descriptive, wdyt?

Reply:
Updated